misc/experimental/rust/celix/src/bundle_context.rs

```diff
@@ -0,0 +1,669 @@
+            many,
+            service_name: type_name::<T>().to_string(),
+            filter: "".to_string(),
+            callback: None,
+        }
+    }
+
+    pub fn with_callback(&mut self, closure: Box<dyn Fn(&T)>)-> &mut Self {
+        self.callback = Some(closure);
+        self
+    }
+
+    pub fn with_service_name(&mut self, name: &str) -> &mut Self {
+        self.service_name = name.to_string();
+        self
+    }
+
+    pub fn with_filter(&mut self, filter: &str) -> &mut Self {
+        self.filter = filter.to_string();
+        self
+    }
+
+    fn validate(&self) -> Result<(), Error> {
+        if self.service_name.is_empty() {
+            return Err(Error::BundleException);
+        }
+        Ok(())
+    }
+
+    unsafe extern "C" fn use_service_c_callback(handle: *mut c_void, svc: *mut c_void) {
+        let closure = handle as *const Box<dyn Fn(&T)>;
+        let closure = closure.as_ref().unwrap();
+
+        let typed_svc = svc as *const T;
+        let typed_svc = typed_svc.as_ref().unwrap();
+
+        closure(typed_svc);
+    }
+
+    pub fn build(&mut self) ->  Result<isize, Error> {
+        self.validate()?;
+
+        let c_service_name = std::ffi::CString::new(self.service_name.as_str()).unwrap();
+        let c_filter = std::ffi::CString::new(self.filter.as_str()).unwrap();
+        let c_service_name_ptr: *const i8 = c_service_name.as_ptr();
+        let c_filter_ptr: *const i8 = if self.filter.is_empty() { null_mut()} else {c_filter.as_ptr() };
+
+        let c_closure_ptr = self.callback.as_ref().unwrap() as *const Box<dyn Fn(&T)> as *mut c_void;
+
+        let opts = celix_service_use_options_t {
+            filter: celix_service_filter_options_t {
+                serviceName: c_service_name_ptr,
+                versionRange: null_mut(),
+                filter: c_filter_ptr,
+                serviceLanguage: null_mut(),
+                ignoreServiceLanguage: false,
+            },
+            waitTimeoutInSeconds: 0.0,
+            callbackHandle: c_closure_ptr,
+            use_: Some(Self::use_service_c_callback),
+            useWithProperties: None,
+            useWithOwner: None,
+            flags: 0,
+        };
+
+        if self.many {
+            unsafe {
+                let count = celix_bundleContext_useServicesWithOptions(self.ctx.get_c_bundle_context(), &opts);
+                Ok(count as isize)
+            }
+        } else {
+            unsafe {
+                let called = celix_bundleContext_useServiceWithOptions(self.ctx.get_c_bundle_context(), &opts);
+                if called {
+                    Ok(1)
+                } else {
+                    Ok(0)
+                }
+            }
+        }
+    }
+}
+
+struct ServiceTrackerCallbacks<T> {
+    set_callback: Option<Box<dyn Fn(Option<&T>)>>,
+    add_callback: Option<Box<dyn Fn(&T)>>,
+    remove_callback: Option<Box<dyn Fn(&T)>>,
+}
+
+pub struct ServiceTracker<T> {
+    ctx: Arc<BundleContext>,
```

Comment:
This does not feel right, since `BundleContext` has longer lifetime than any tracker associated with it.
This weirdness leads to unnatural `BundleContext::get_self`.

Reply:
Lifetime Order: BundleContext > Activator > Loggerhelper > ServiceTracker ...
If we can enforce such order using Rust type system, then lots of weirdness will disappear.